Support bidirectional queue system

alanjian85 opened the pull request:
Use the newly introduced bidirectional system calls setup_queue and submit_queue to avoid crashing when launching the demo.

jserv reviewed (changes requested):
Remove `.clang-format` and minimize the necessary changes. That is, you shall follow the original coding style (despite being messy) and make functional modifications.

alanjian85 committed:
Support bidirectional queue system

Use the newly introduced bidirectional system calls setup_queue and
submit_queue to avoid crashing when launching the demo.

jserv commented on port/boards/rv32emu/main.c:
```diff
@@ -74,6 +109,15 @@ void qembd_udelay(uint32_t us)
+    submission_queue.base = base + sizeof(event_t) * queues_capacity;
+    register int a0 asm("a0") = (uintptr_t) base;
+    register int a1 asm("a1") = queues_capacity;
+    register int a2 asm("a2") = (uintptr_t) & event_count;
```
Rewrite as following:
```c
register int a2 asm("a2") = (uintptr_t) (&event_count);
```

Sometimes, `clang-format` was stupid.

jserv commented on port/boards/rv32emu/main.c:
```diff
@@ -84,14 +128,18 @@ void *qembd_allocmain(size_t size)
-	register int a0 asm("a0") = (uintptr_t) &event;
-	register int a7 asm("a7") = 0xc0de;
-	asm volatile("scall" : "+r"(a0) : "r"(a7));
+    if (event_count <= 0)
```
Follow the original coding style. e.g., using tabs.

jserv commented on port/boards/rv32emu/main.c:
```diff
@@ -84,14 +128,18 @@ void *qembd_allocmain(size_t size)
 	}
-	return a0;
+	
+    return 1;
```
Why did you return `1` here?

alanjian85 replied:
> Why did you return `1` here?

I return an integer here to inform the caller whether or not there are any events in the queue. The return value of `poll_event()` is used in the `qembd_dequeue_key_event()` to determine whether the loop should continue or not.

jserv commented:
This patch looks great! However, I encountered the following error while attempting to quit:
```
You got the Grenade Launcher
2022-10-18 23:19:17.957 rv32emu[61973:4131715] TSM AdjustCapsLockLEDForKeyTransitionHandling - _ISSetPhysicalKeyboardCapsLockLED Inhibit
/bin/sh: line 1: 61973 Segmentation fault: 11  ../build/rv32emu quake.elf
```

Then, it failed to launch again:
```
execing quake.rc
execing default.cfg
Unknown command "volume"
FindFile: .//id1/config.cfg
/bin/sh: line 1: 62045 Segmentation fault: 11  ../build/rv32emu quake.elf
```

alanjian85 commented:
> This patch looks great! However, I encountered the following error while attempting to quit:
> 
> ```
> You got the Grenade Launcher
> 2022-10-18 23:19:17.957 rv32emu[61973:4131715] TSM AdjustCapsLockLEDForKeyTransitionHandling - _ISSetPhysicalKeyboardCapsLockLED Inhibit
> /bin/sh: line 1: 61973 Segmentation fault: 11  ../build/rv32emu quake.elf
> ```
> 
> Then, it failed to launch again:
> 
> ```
> execing quake.rc
> execing default.cfg
> Unknown command "volume"
> FindFile: .//id1/config.cfg
> /bin/sh: line 1: 62045 Segmentation fault: 11  ../build/rv32emu quake.elf
> ```

I found a critical bug in rv32emu's implementation of red-black tree; I opened a pull-request to fix that bug, so I believe this pull-request is fine and can be safely merged.
